### PR TITLE
tree-sitter: 0.15.7 -> 0.16.4

### LINKED ISSUE #75863: tree-sitter out of date
**Describe the bug**
 tree-sitter on nixos-unstable doesn't build.
So I tried bumping the package but ended up with another issue:

```
---- tests::highlight_test::test_highlighting_ejs stdout ----
thread 'tests::highlight_test::test_highlighting_ejs' panicked at 'called `Result::unwrap()` on an `Err` value: Error(["/build/source/target/scratch/html.so: invalid ELF header", "Error opening dynamic library \"/build/source/target/scratch/html.so\""])', src/libcore/result.rs:1165:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

```
May need a patchelf ?

**To Reproduce**
nix build tree-sitter -f channel:nixos-unstable

**Expected behavior**

**Metadata**
```
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.84, NixOS, 20.03.git.2bb799f (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.1`
 - channels(teto): `"nixos-18.09.tar.gz-18.09"`
 - channels(root): `""`
 - nixpkgs: `/home/teto/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: tree-sitter
# a list of nixos modules affected by the problem
module:
```
CC @Profpatsch as per  https://github.com/NixOS/nixpkgs/pull/66758

### PR DESCRIPTION
**Pull request by bhipple**
Includes some bugfixes/cleanups to the scripts and packaging, a run of the
updater, a bump of the version, an upgrade to the newer cargo fetcher in #79975,
and gets the web assembly portion to compile successfully.

Fixes #75863

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bhipple** [comment]
@GrahamcOfBorg build tree-sitter

File: pkgs/development/tools/parsing/tree-sitter/update.nix
Comment (bhipple) [inline pkgs/development/tools/parsing/tree-sitter/update.nix]: This never worked previously, but its failure just prints a warning :)

File: pkgs/development/tools/parsing/tree-sitter/default.nix
Comment (bhipple) [inline pkgs/development/tools/parsing/tree-sitter/default.nix]: It's somewhat flaky (on master as well).  Every run has a different number of failures; here's my best run so far:

```
failures:
    tests::corpus_test::test_feature_corpus_files
    tests::corpus_test::test_real_language_corpus_files
    tests::parser_test::test_parsing_with_timeout_and_no_completion
    tests::query_test::test_query_captures_with_duplicates
    tests::query_test::test_query_matches_in_language_with_simple_aliases

test result: FAILED. 128 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out
```

File: pkgs/development/tools/parsing/tree-sitter/default.nix
Comment (bhipple) [inline pkgs/development/tools/parsing/tree-sitter/default.nix]: We now support web assembly and can remove our old patch (which didn't apply cleanly) and vendoring hacks.  Someone who wants to go really above and beyond could also get the JS minify step working, but I'm satisfied taking it to this point.

**Review by Ma27 — APPROVED** [review]

**Comment by Profpatsch** [comment]
@Ma27 @bhipple would have been great if I had been pinged as maintainer before merging.

File: pkgs/development/tools/parsing/tree-sitter/default.nix
Comment (Profpatsch) [reply pkgs/development/tools/parsing/tree-sitter/default.nix]: Let’s add an option to make the web-ui optional.

**Comment by Profpatsch** [comment]
Thanks! Do you want to add yourself as a maintainer?

**Comment by Ma27** [comment]
Backported the update slightly altered as 57fafc08f95e2519dce88a259761d7af69e5b519 to fix the build on release-20.03 as well.

**Pull request merged**
